edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ScopeIdentitiesHandler.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity.Service;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class ScopeIdentitiesHandler : IMessageProducer
+    {
+        const string Topic = "$internal/identities";
+        readonly IDeviceScopeIdentitiesCache deviceScopeIdentitiesCache;
+        readonly AtomicBoolean connected = new AtomicBoolean(false);
+        IMqttBrokerConnector connector;
+        IList<BrokerServiceIdentity> lastBrokerServiceIdentityUpdate = new List<BrokerServiceIdentity>();
+
+        public ScopeIdentitiesHandler(IDeviceScopeIdentitiesCache deviceScopeIdentitiesCache)
+        {
+            this.deviceScopeIdentitiesCache = deviceScopeIdentitiesCache;
+            this.deviceScopeIdentitiesCache.ServiceIdentitiesUpdated +=
+                async (sender, serviceIdentities) => await this.ServiceIdentitiesUpdated(serviceIdentities);
+        }
+
+        public void SetConnector(IMqttBrokerConnector connector)
+        {
+            this.connector = connector;
+            this.connector.OnConnected += async (sender, args) => await this.OnConnect();
+        }
+
+        async Task OnConnect()
+        {
+            this.connected.Set(true);
+            await this.PublishBrokerServiceIdentities(this.lastBrokerServiceIdentityUpdate);
+        }
+
+        async Task ServiceIdentitiesUpdated(IList<ServiceIdentity> serviceIdentities)
+        {
+            IList<BrokerServiceIdentity> brokerServiceIdentities =
+                    await Task.WhenAll(
```

Comment:
I'm not sure if running `GetAuthChain` in parallel will provide a benefit. AFAIK, `ServiceIdentityTree` implementation uses nodeLock inside, so each call will be exclusive anyways. 

If you want to make it fast, maybe we should add a overloaded method to `IServiceIdentityHierarchy` that accepts list of ids and returns the auth chains.

@richma-ms do you have any thoughts on that?

Reply:
I changed it to just be a loop that calls GetAuthChain on every Id. We can make it marginally faster if we want to add that extra method to IServiceIdentityHierarchy

Reply:
I don't think it'll make a noticeable difference, authchain is calculcated and cached on identity insertion/updation, so GetAuthChain is just a constant-time look up.

---
In reply to: [477477814](https://github.com/Azure/iotedge/pull/3439#discussion_r477477814) [](ancestors = 477477814)